build.rs

```diff
@@ -121,35 +100,25 @@ static SYSCONFIG_VALUES: [&str; 1] = [
 
 /// Attempts to parse the header at the given path, returning a map of definitions to their values.
 /// Each entry in the map directly corresponds to a `#define` in the given header.
-fn parse_header_defines<P: AsRef<Path>>(header_path: P) -> Result<HashMap<String, String>, String> {
+fn parse_header_defines(header_path: impl AsRef<Path>) -> Result<HashMap<String, String>> {
     // This regex picks apart a C style, single line `#define` statement into an identifier and a
     // value. e.g. for the line `#define Py_DEBUG 1`, this regex will capture `Py_DEBUG` into
     // `ident` and `1` into `value`.
-    let define_regex =
-        Regex::new(r"^\s*#define\s+(?P<ident>[a-zA-Z0-9_]+)\s+(?P<value>.+)\s*$").unwrap();
+    let define_regex = Regex::new(r"^\s*#define\s+(?P<ident>[a-zA-Z0-9_]+)\s+(?P<value>.+)\s*$")?;
 
-    let header_file = File::open(header_path.as_ref()).map_err(|e| e.to_string())?;
+    let header_file = File::open(header_path.as_ref())?;
     let header_reader = BufReader::new(&header_file);
 
-    let definitions = header_reader
-        .lines()
-        .filter_map(|maybe_line| {
-            let line = maybe_line.unwrap_or_else(|err| {
-                panic!("failed to read {}: {}", header_path.as_ref().display(), err);
-            });
-            let captures = define_regex.captures(&line)?;
-
-            if captures.name("ident").is_some() && captures.name("value").is_some() {
-                Some((
-                    captures.name("ident").unwrap().as_str().to_owned(),
-                    captures.name("value").unwrap().as_str().to_owned(),
-                ))
-            } else {
-                None
-            }
-        })
-        .collect();
-
+    let mut definitions = HashMap::new();
+    let tostr = |r: regex::Match<'_>| r.as_str().to_string();
+    for maybe_line in header_reader.lines() {
+        if let Some(captures) = define_regex.captures(&maybe_line?) {
+            match (captures.name("ident"), captures.name("value")) {
```

Comment:
if the regex can't match without capturing the groups, you can use `captures["ident"]` etc which already gives you a `&str`.

Reply:
Thanks, but looks we need `Option` here.